### PR TITLE
Supporter helper tests

### PR DESCRIPTION
**Pull request by martinfrancois**
Finally implemented tests for the SupporterHelper class and also for the Supporter class. Refactored SupporterHelper class to be more easily testable and more reliable overall, by improving catching of extraordinary cases.
This increases code coverage to 97% for SupporterHelper and 100% for Supporter

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -459,5 +459,13 @@
             <version>4.7.0-5</version>
         </dependency>
 
+        <!-- http://jqno.nl/equalsverifier -->
```
Comment (l-marchesi) [inline pom.xml]: Maybe add why you need this particular dependency in a short sentence. Otherwise it seems unclear to me, why you should need it.

File: src/ch/imedias/rsccfx/model/xml/Supporter.java [hunk 2]
```diff
@@ -161,4 +162,34 @@ public String toString() {
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    Supporter supporter = (Supporter) o;
+    return encrypted == supporter.encrypted
```
Comment (l-marchesi) [inline src/ch/imedias/rsccfx/model/xml/Supporter.java]: it was a bit hard to understand for me, maybe add "this." in front of the fields, to make it clearer?

File: src/ch/imedias/rsccfx/model/xml/SupporterHelper.java [hunk 3]
```diff
@@ -65,28 +68,26 @@ public void saveSupporters(List<Supporter> supporters) {
...
-
-      Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
-      Supporters supporters = (Supporters) jaxbUnmarshaller.unmarshal(file);
-
-      supportersList = supporters.getSupporters();
-    } catch (JAXBException e) {
-      e.printStackTrace();
-    }
-    return supportersList;
+    return getSupportersFromXml(fileToString(file));
```
Comment (l-marchesi) [inline src/ch/imedias/rsccfx/model/xml/SupporterHelper.java:82]: It is a bit unclear to me, why you removed those lines and replaced it with only one?

File: src/ch/imedias/rsccfx/view/SupporterAttributesDialog.java [hunk 4]
```diff
@@ -116,8 +112,6 @@ private void layoutForm() {
     attributePane.add(addressFld, 1, 1);
     attributePane.add(portLbl, 0, 2);
     attributePane.add(portFld, 1, 2);
-    attributePane.add(pictureLbl, 0, 3);
```
Comment (l-marchesi) [inline src/ch/imedias/rsccfx/view/SupporterAttributesDialog.java:119]: Why did you remove pictureLbl and pictureFld? The pull request is about supporterHelperTests...

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    assertEquals(2, actualSupporters.size());
+    assertEquals(supporter1, actualSupporters.get(0));
+    assertEquals(supporter2, actualSupporters.get(1));
+    actualSupporters = null;
+
+    // if preferences contain invalid xml
+    String invalidXml = "<test>this is not a correctly formatted xml</not test>";
+    supporterHelper.setSupportersInPreferences(invalidXml);
+    actualSupporters = supporterHelper.loadSupporters();
+    verifyDefaultSupporters(actualSupporters);
```
Comment (l-marchesi) [inline test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java:92]: Since you're loading a wrong XML file... shouldn't this test then fail instead of pass?

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    assertEquals(supportersXml, supporterHelper.getSupportersXmlFromPreferences());
+  }
+
+  /**
+   * Test for {@link SupporterHelper#getDefaultSupporters()}.
+   */
+  @Test
+  public void testGetDefaultSupporters() throws Exception {
+    List<Supporter> actualSupporters = supporterHelper.getDefaultSupporters();
+    verifyDefaultSupporters(actualSupporters);
```
Comment (l-marchesi) [inline test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java:112]: For me it is a bit weird, that you call a method several times, which is basically yet another test. Personally I would have gone with a method who returns true or false (verifyDefaultSupporters) to make it clearer.

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  /**
+   * Verifies that the supporters list is actually the list of default supporters.
+   * Not marked with a @Test annotation because it is indirectly called in other tests.
+   */
+  private void verifyDefaultSupporters(List<Supporter> actualSupporters) {
```
Comment (l-marchesi) [inline test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java]: If you're sticking with the non-test-but-somehow-test-anyway thing, I would move this at the end or at the beginning of the file. It is quite easy to not get that "@test" is missing while scrolling through

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    assertEquals(supporter2, actualSupporters.get(1));
+    assertEquals(supporterLast, actualSupporters.get(DEFAULT_SUPPORTER_SIZE - 1));
+  }
+
+  /**
+   * Tests what happens when the default supporters path is either non-existent or null.
+   */
+  @Test
+  public void testInvalidDefaultSupportersPath() {
+    // should not throw any exceptions but cause logger entries
```
Comment (l-marchesi) [inline test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java:120]: What happens if it throws an exception for some reason?

File: test/ch/imedias/rsccfx/model/xml/SupporterTest.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    String testDescription = "Test Description";
+    supporter.setDescription(testDescription);
+    assertEquals(testDescription, supporter.toString());
+  }
+
+  /**
+   * Test for {@link Supporter#equals(Object)} and {@link Supporter#hashCode()}.
+   */
+  @Test
+  public void testEqualsContract() {
```
Comment (l-marchesi) [inline test/ch/imedias/rsccfx/model/xml/SupporterTest.java:44]: It is unclear for me, why this test is needed. It says only that it has nothing to do with HashMap or HashSet, but what is it correlated to?
And which contract do you mean? (testEqualsContract) --> the contract with the customer?

**Review by l-marchesi — CHANGES_REQUESTED** [review]

File: src/ch/imedias/rsccfx/model/xml/SupporterHelper.java (hunk 3 above)
Comment (martinfrancois) [reply src/ch/imedias/rsccfx/model/xml/SupporterHelper.java:82]: Since the code was essentially duplicated before with only minor changes, I decided to extract the parts that were different to "fileToString()" and use the similar method to avoid duplication. Also, this allows for easier and more thorough testing.

File: src/ch/imedias/rsccfx/view/SupporterAttributesDialog.java (hunk 4 above)
Comment (martinfrancois) [reply src/ch/imedias/rsccfx/view/SupporterAttributesDialog.java:119]: I forgot to mention that in the PR, @Kennox removed it since we decided not to implement the picture fields

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java (hunk 5 above)
Comment (martinfrancois) [reply test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java:92]: The implementation is as follows, that if anything unexpected happens during loading of the xml files, it should resort to loading the default supporters instead. So it would fail if the default supporters were not loaded, hence why I was testing if they were loaded.

File: test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java (hunk 8 above)
Comment (martinfrancois) [reply test/ch/imedias/rsccfx/model/xml/SupporterHelperTest.java:120]: then the test will fail and we will know that something is wrong?

File: test/ch/imedias/rsccfx/model/xml/SupporterTest.java (hunk 9 above)
Comment (martinfrancois) [reply test/ch/imedias/rsccfx/model/xml/SupporterTest.java:44]: http://www.programcreek.com/2011/07/java-equals-and-hashcode-contract/

**Review by rmosimann — APPROVED** [review]
Nice!
